podvm/Dockerfile.podvm

```diff
@@ -9,55 +9,105 @@ ARG BUILDER_IMG="quay.io/confidential-containers/podvm-builder-ubuntu"
+
+ARG RHEL_IMAGE_URL="/tmp/rhel.qcow2"
+ARG RHEL_IMAGE_CHECKSUM
+
+
+FROM ${BINARIES_IMG} AS binaries
+
+FROM scratch AS hostkeys
+
+# Copy HOST KEY documents to podvm/files folder for Secure Execution
+COPY podvm/files/HKD-*.crt /src/cloud-api-adaptor/podvm/files/
+
+FROM ${EXTRAFILES_IMG:-scratch} AS extrafiles
+
+FROM ${BUILDER_IMG} AS builder-common
+
+ARG ARCH
 ENV ARCH ${ARCH}
-ENV UEFI ${UEFI}
-ENV SE_BOOT ${SE_BOOT}
 
-RUN if [ -n "${CAA_SRC}" ]; then \
-      rm -rf cloud-api-adaptor && \
-      git clone ${CAA_SRC} cloud-api-adaptor;\
-    fi && \
-    if [ -n "${CAA_SRC_REF}" ]; then \
-      cd cloud-api-adaptor && \
-      git fetch origin ${CAA_SRC_REF} && \
-      git checkout FETCH_HEAD -b ${CAA_SRC_REF} ;\
-    fi
 # Installs add-ons for foreign target, if required
 RUN ./cloud-api-adaptor/podvm/hack/cross-build-extras.sh
 
-# Defaults to Ubuntu Focal amd64 release image. These variables can be overriden as needed
-ARG UBUNTU_IMAGE_URL=https://cloud-images.ubuntu.com/releases/focal/release-20230107/ubuntu-20.04-server-cloudimg-amd64.img
-ARG UBUNTU_IMAGE_CHECKSUM=3895e38566e5c2c019f5c6f825ab7570ee34dac6b9142fab0c7e5a78084c4280
+ARG UBUNTU_IMAGE_URL
+ARG UBUNTU_IMAGE_CHECKSUM
+ARG CENTOS_IMAGE_URL
+ARG CENTOS_IMAGE_CHECKSUM
+ARG RHEL_IMAGE_URL
+ARG RHEL_IMAGE_CHECKSUM
 
 ENV UBUNTU_IMAGE_URL ${UBUNTU_IMAGE_URL}
 ENV UBUNTU_IMAGE_CHECKSUM ${UBUNTU_IMAGE_CHECKSUM}
+ENV CENTOS_IMAGE_URL ${CENTOS_IMAGE_URL}
+ENV CENTOS_IMAGE_CHECKSUM ${CENTOS_IMAGE_CHECKSUM}
+ENV RHEL_IMAGE_URL ${RHEL_IMAGE_URL}
+ENV RHEL_IMAGE_CHECKSUM ${RHEL_IMAGE_CHECKSUM}
 
 # Copy the binaries to podvm/files folder
-COPY --from=podvm_binaries /podvm-binaries.tar.gz /src/cloud-api-adaptor/podvm/files
+COPY --from=binaries /podvm-binaries.tar.gz /src/cloud-api-adaptor/podvm/files
 RUN tar xvf /src/cloud-api-adaptor/podvm/files/podvm-binaries.tar.gz -C /src/cloud-api-adaptor/podvm/files
 
 # Copy the pause_bundle to podvm/files folder
-COPY --from=podvm_binaries /pause-bundle.tar.gz /src/cloud-api-adaptor/podvm/files
+COPY --from=binaries /pause-bundle.tar.gz /src/cloud-api-adaptor/podvm/files
 RUN tar xvf /src/cloud-api-adaptor/podvm/files/pause-bundle.tar.gz -C /src/cloud-api-adaptor/podvm/files
 
-# Copy HOST KEY documents to podvm/files folder for Secure Execution
-# Use one files/.dummy.crt file at here to avoid `COPY failed: no source files were specified` for non-SE build
-COPY files/*.crt /src/cloud-api-adaptor/podvm/files/
+FROM builder-common AS builder-remote
+
+RUN if [ -n "${CAA_SRC}" ]; then \
+      rm -rf cloud-api-adaptor && \
+      git clone ${CAA_SRC} cloud-api-adaptor;\
+    fi && \
+    if [ -n "${CAA_SRC_REF}" ]; then \
+      cd cloud-api-adaptor && \
+      git fetch origin ${CAA_SRC_REF} && \
+      git checkout ${CAA_SRC_REF} ;\
+    fi
+
+FROM builder-common AS builder-local
+
+RUN rm -fr /src/cloud-api-adaptor
+COPY . /src/cloud-api-adaptor
+
+FROM builder-${SOURCE_FROM} AS builder
```

Comment:
This looks like a good approach to unify the builds. We'll need to ensure we use buildkit so that only the specific build stage can be built as needed. Ref: https://docs.docker.com/build/building/multi-stage/#differences-between-legacy-builder-and-buildkit